onnxruntime/core/providers/cuda/math/clip_impl.cu

```diff
@@ -19,13 +19,23 @@ void ClipImpl(cudaStream_t stream, const T* input_data, T* output_data, const T*
   typedef typename ToCudaType<T>::MappedType CudaT;
 
   int blocksPerGrid = (int)(ceil(static_cast<float>(count) / GridDim::maxThreadsPerBlock));
-  _Clip<CudaT><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(reinterpret_cast<const CudaT*>(input_data),
-                                                                  reinterpret_cast<CudaT*>(output_data),
-                                                                  reinterpret_cast<const CudaT*>(min),
-                                                                  reinterpret_cast<const CudaT*>(max),
-                                                                  *reinterpret_cast<CudaT*>(&min_default),
-                                                                  *reinterpret_cast<CudaT*>(&max_default),
-                                                                  count);
+  union ConstAlias {
+    const T *t;
+    const CudaT *cudaT;
+    const_alias(const T* _t) { t = _t;}
+  };
+  union Alias {
+    T *t;
+    CudaT *cudaT;
+    alias(T* _t) { t = _t;}
+  };
+  _Clip<CudaT><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(((union ConstAlias)input_data).cudaT,
+                                                                          ((union Alias)output_data).cudaT,
+                                                                          ((union ConstAlias)min).cudaT,
+                                                                          ((union ConstAlias)max).cudaT,
+                                                                          *((union Alias)&min_default).cudaT,
+                                                                          *((union Alias)&max_default).cudaT,
+                                                                          count);
```

Comment:
Why do we need this change?

Reply:
The previous version was using reinterprete_cast to convert a pointer to type T to a pointer to type CudaT, in this template function. It seems, the nvcc compiler we are using on the build server is stricter, somehow applying stricter aliasing rules. 
Type Punning means interpreting bits differently from the original declaration. The compiler thew error because it is finding aliasing violations. The GXX complier flag -Wno-strict-aliasing is provided to suppress the error. 
However NVCC does not support the flag. There is no compiler option -Wno-strict-aliasing in the document.
https://docs.nvidia.com/pdf/CUDA_Compiler_Driver_NVCC.pdf
So, relatively better way in this case than reinterpret_cast is to create union.
The compiler is happy!

Reply:
thanks for the explanation, good to learn as well.